src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProvider.java

```diff
@@ -0,0 +1,296 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.semanticTokens.viewProvider;
+
+import com.intellij.lang.Language;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.*;
+import com.intellij.psi.util.CachedValueProvider;
+import com.intellij.psi.util.CachedValuesManager;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.client.features.EditorBehaviorFeature;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link FileViewProvider} for an LSP-backed file where elements are derived dynamically from reported semantic tokens.
+ */
+@ApiStatus.Internal
+public class LSPSemanticTokensFileViewProvider extends SingleRootFileViewProvider {
+
+    /**
+     * Creates a new file view provider based on language.
+     *
+     * @param psiManager         the PSI manager
+     * @param virtualFile        the virtual file
+     * @param eventSystemEnabled whether or not the event system is enabled
+     * @param language           the language which should always be TextMate
+     */
+    LSPSemanticTokensFileViewProvider(@NotNull PsiManager psiManager,
+                                      @NotNull VirtualFile virtualFile,
+                                      boolean eventSystemEnabled,
+                                      @NotNull Language language) {
+        super(psiManager, virtualFile, eventSystemEnabled, language);
+    }
+
+    /**
+     * Creates a new file view provider based on file type.
+     *
+     * @param psiManager         the PSI manager
+     * @param virtualFile        the virtual file
+     * @param eventSystemEnabled whether or not the event system is enabled
+     */
+    LSPSemanticTokensFileViewProvider(@NotNull PsiManager psiManager,
```

Comment:
And this creates a file view provider using the file type when we don't have a language.